server/src/main/java/org/elasticsearch/index/shard/GlobalCheckpointListeners.java

```diff
@@ -0,0 +1,161 @@
+    private final Logger logger;
+
+    /**
+     * Construct a global checkpoint listeners collection.
+     *
+     * @param shardId  the shard ID on which global checkpoint updates can be listened to
+     * @param executor the executor for listener notifications
+     * @param logger   a shard-level logger
+     */
+    GlobalCheckpointListeners(
+            final ShardId shardId,
+            final Executor executor,
+            final Logger logger) {
+        this.shardId = Objects.requireNonNull(shardId);
+        this.executor = Objects.requireNonNull(executor);
+        this.logger = Objects.requireNonNull(logger);
+    }
+
+    /**
+     * Add a global checkpoint listener. If the global checkpoint is above the current global checkpoint known to the listener then the
+     * listener will be asynchronously notified on the executor used to construct this collection of global checkpoint listeners. If the
+     * shard is closed then the listener will be asynchronously notified on the executor used to construct this collection of global
+     * checkpoint listeners.
+     *
+     * @param currentGlobalCheckpoint the current global checkpoint known to the listener
+     * @param listener                the listener
+     */
+    synchronized void add(final long currentGlobalCheckpoint, final GlobalCheckpointListener listener) {
+        if (closed) {
+            executor.execute(() -> listener.accept(UNASSIGNED_SEQ_NO, new IndexShardClosedException(shardId)));
+        }
+        if (lastKnownGlobalCheckpoint > currentGlobalCheckpoint) {
+            // notify directly
+            executor.execute(() -> notifyListener(listener, lastKnownGlobalCheckpoint, null));
+        } else {
+            if (listeners == null) {
+                listeners = new ArrayList<>();
+            }
+            listeners.add(listener);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        synchronized (this) {
+            closed = true;
+        }
+        notifyListeners(UNASSIGNED_SEQ_NO, new IndexShardClosedException(shardId));
+    }
+
+    /**
+     * Invoke to notify all registered listeners of an updated global checkpoint.
+     *
+     * @param globalCheckpoint the updated global checkpoint
+     */
+    void globalCheckpointUpdated(final long globalCheckpoint) {
+        assert globalCheckpoint >= NO_OPS_PERFORMED;
+        synchronized (this) {
+            lastKnownGlobalCheckpoint = globalCheckpoint;
+        }
+        notifyListeners(globalCheckpoint, null);
+    }
+
+    private void notifyListeners(final long globalCheckpoint, final IndexShardClosedException e) {
+        assert (globalCheckpoint == UNASSIGNED_SEQ_NO && e != null) || (globalCheckpoint >= NO_OPS_PERFORMED && e == null);
+        if (listeners != null) {
+            final List<GlobalCheckpointListener> currentListeners;
+            synchronized (this) {
+                currentListeners = listeners;
+                listeners = null;
```

Comment:
This looks as if the listeners are only notified once and then need to reregister if they want more events? I can see why this kind of behavior makes sense for the refresh listeners, with refresh being an expensive operation that ensures that all events registered before the refresh will now see the changes they're waiting for. With global checkpoints, it's less clear to me, as they can be potentially updated many many times per second, so wouldn't you want to stay registered to receive events. If not, will this lead to a storm of reregister events? I think I need to better understand the integration point here, i.e., how this API will be used.

Reply:
@ywelsch See #32651. The intended usage is in CCR where a remote cluster will, when the remote cluster is fully caught up, (remotely) attach a single-use listener to the local cluster for the next global checkpoint change. When the global checkpoint is updated, the listener will be invoked which will return a response to the remote cluster that will act as letting the remote cluster know that there are now additional changes to be fetched.

Reply:
Ok, given this behavior, I wonder if we can make the listener interface simpler (same as Boaz's concern). As the listeners are only notified once and then need to reregister if they want more events, I wonder if it's simpler to just signal an UNASSIGNED_SEQ_NO (or the lastKnownGlobalCheckpoint) on a closing and then have the caller fail on a repeated call to the add method (by throwing directly the exception in that method, not relaying it to the listener). This means that the listener can remain a functional interface (but just a LongConsumer). WDYT?

Reply:
@ywelsch Personally I do not buy the argument that

```java
        @Override
        protected void asyncShardOperation(
                final Request request, final ShardId shardId, final ActionListener<Response> listener) throws IOException {
            final IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
            final IndexShard indexShard = indexService.getShard(shardId.id());
            indexShard.addGlobalCheckpointListener(
                    request.getGlobalCheckpoint(),
                    (g, e) -> {
                        if (g != UNASSIGNED_SEQ_NO) {
                            listener.onResponse(new Response(g));
                        } else {
                            listener.onFailure(e);
                        }
            });
        }
```

is less clean than

```java
        @Override
        protected void asyncShardOperation(
                final Request request, final ShardId shardId, final ActionListener<Response> listener) throws IOException {
            final IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
            final IndexShard indexShard = indexService.getShard(shardId.id());
            indexShard.addGlobalCheckpointListener(
                    request.getGlobalCheckpoint(),
                    g -> {
                        if (g != UNASSIGNED_SEQ_NO) {
                            listener.onResponse(new Response(g));
                        } else {
                            listener.onFailure(new IndexShardClosedException(shardId));
                        }
            });
        }
```

We still have to have a check in one form or the other whether or not closing has been signaled to us, so there's always going to be an `if` check. At least, that is what I read the [comment](
https://github.com/elastic/elasticsearch/pull/32696/files#r209377535) from @bleskes as arguing:

> The down side is of course that people wouldn't be able to pass a method references, but the method won't need to start with if (e != null) etc.

In fact, I would argue the approach I have taken is cleaner as it's the shard telling us that we are closed rather than it being signaled indirectly through the value of the global checkpoint passed in the callback. Sure the actual interface is simpler but I prefer the explicit approach. So I think we should either stick with what I have, or have an `onClosed` callback and lose the ability for the interface to be a functional interface.

Regarding throwing on attempting to register a listener on a closed shard, that was my [preferred](
https://github.com/elastic/elasticsearch/pull/32696#discussion_r209378359) approach too but @bleskes thought otherwise.

Reply:
There also still the option of adding an onClosed method to the interface with a default NOOP implementation. It will remain a functional interface, and if most tests don't care about the shard closed case, they can treat it like a functional interface. I'll leave the decision to you.

Reply:
It would not be a functional interface for the purposes of the production use-case for this that I have in mind (we have to handle the shard closed event). Thanks, I will leave as-is.